File: project/Validate.hs
-- Navraj Narula

module Exhaustive where

import Error
import AbstractSyntax
import Parse
import TypeCheck
import Interpret
import Compile
import Machine

type Height = Integer
type Quantity = Integer

--LOOK AT NOTES ON LIST COMPREHENSIONS!!!!!!!!!!!

class Exhaustive a where
  exhaustive :: Integer -> [a]

instance Exhaustive Stmt where
  exhaustive 0 = [] 
  exhaustive x = if == 1 then [(End)] else [Print ex st | ex <- ((exhaustive (x-1)::[Exp]), st <- ((exhaustive (x-1))::[Stmt])]
  	++[Assign X ex st | ex <- ((exhaustive (x-1)):: [Exp]), s <- ((exhaustive (x-1)):: [Stmt])]
  	++[Assign Y ex st | ex <- ((exhaustive (x-1)):: [Exp]), s <- ((exhaustive (x-1)):: [Stmt])]

instance Exhaustive Exp where
  exhaustive 0 = []
  exhaustive x = if x == 1 then [(Value True), (Value False), (Variable X), (Variable Y)] else
  	[And e1 e2 | e1 <- exhaustive(x-1), e2 <- exhaustive(x-1)]
  	++[Or e1 e2 | e1 <- exhaustive(x-1), e2 <- exhaustive(x-1)]


take' :: Integer -> [a] -> [a]
take' n (x:xs) = x:(take' (n-1) xs)
take' 0 _      = []
take' _ _      = []

validate :: Height -> Quantity -> Bool
validate n k = False -- Complete for Problem 4, part (c).

complete :: String -> ErrorOr Buffer
complete _ = TypeError "Complete for Problem 4, part (d)."

--eof

File: midterm/validate.py
#####################################################################
#
# CAS CS 320, Fall 2015
# Midterm (skeleton code)
# validate.py
#
#  ****************************************************************
#  *************** Modify this file for Problem #5. ***************
#  ****************************************************************
#
# Modified by: Navraj Narula

exec(open('analyze.py').read())
exec(open('interpret.py').read())
exec(open('compile.py').read())

def convertValue(v):
    if type(v) == Leaf:
        if v == 'True':
            return 1
        if v == 'False':
            return 0
    if type(v) == Node:
        for label in v:
            return 1 + max([convertValue(child) for child in v[label]])

# Converts an output (a list of values) from the
# value representation to the machine representation
def convert(o):
    return [convertValue(v) for v in o]

def expressions(n):
    if n <= 0:
        []
    elif n == 1:
         # Add all base case(s) for Problem #5.
         # The only possible element variable name is 'a.'
        return ['True','False',{'Element': [{'Variable': ['a']}, {'Number': [1]}]},{'Element': [{'Variable': ['a']}, {'Number': [1]}]},{'Element': [{'Variable': ['a']}, {'Number': [1]}]}]
    else:
        # Add recursive case(s) for Problem #5.
        exp = expressions(n-1)
        expVal = []
        return exp + expVal

def programs(n):
    if n <= 0:
        []
    elif n == 1:
        return ['End'] 
    else:
        ps = programs(n-1)
        es = expressions(n-1)
        psN = []
        # Add more nodes to psN for Problem #5.
        psN = psN + [{'Assign':[{'Variable':['a']}, exp, exp, exp, pro]} for pro in ps for exp in es]
        psN = psN + [{'Print': [exp, pro]} for exp in es for pro in ps]
        psN = psN + [{'Loop': [{'Variable':['x']},pro1, pro2]} for pro1 in ps for pro2 in ps]
        return ps + psN
   
# Compute the formula that defines correct behavior for the
# compiler for all program parse trees of depth at most k.
# Any outputs indicate that the behavior of the compiled
# program does not match the behavior of the interpreted
# program.

def exhaustive(k):
    for p in programs(4):
        try:
            if simulate(compile(p)) != convert(execute(p)):
                print('\nIncorrect behavior on: ' + str(p))
        except:
            print('\nError on: ' + str(p))

#eof
